asllib/Typing.ml

```diff
@@ -3625,81 +3625,133 @@ module Annotate (C : ANNOTATE_CONFIG) : S = struct
     | V0 -> annotate_func_sig_v0 ~loc genv func_sig
     | V1 -> annotate_func_sig_v1 ~loc genv func_sig
 
-  module ControlFlow : sig
-    val check_stmt_returns_or_throws : identifier -> stmt_desc annotated -> prop
-    (** [check_stmt_interrupts name body] checks that the function named [name]
-        with the statement body [body] either: returns a value, throws an
-        exception, or calls [Unreachable()].
-        It executes only when [C.control_flow_analysis] is [true]. *)
+  (** A module for checking that a subprogram body satisfies
+      control flow requirements.
+  *)
+  module ControlFlowAnalysis : sig
+    val check_control_flow : env -> func -> stmt -> unit
   end = struct
-    (** Possible Control-Flow actions of a statement. *)
-    type t =
-      | Interrupt  (** Throwing an exception or returning a value. *)
-      | AssertedNotInterrupt
-          (** Assert that this control-flow path is unused. *)
-      | MayNotInterrupt
-          (** Among all control-flow path in a statement, there is one that
-              will not throw an exception nor return a value. *)
-
-    (* Begin ControlFlowSeq *)
-
-    (** Sequencial combination of two control flows. *)
-    let seq t1 t2 =
-      if t1 = MayNotInterrupt then t2 else t1 |: TypingRule.ControlFlowSeq
-    (* End *)
-
-    (* Begin ControlFlowJoin *)
-
-    (** [join t1 t2] corresponds to the parallel combination of [t1] and [t2].
-    More precisely, it is the maximal element in the ordering
-    AssertedNotInterrupt < Interrupt < MayNotInterrupt
+    (** Abstract values representing the possible configurations
+        resulting from evaluating statements.
     *)
-    let join t1 t2 =
-      match (t1, t2) with
-      | MayNotInterrupt, _ | _, MayNotInterrupt ->
-          MayNotInterrupt |: TypingRule.ControlFlowJoin
-      | AssertedNotInterrupt, t | t, AssertedNotInterrupt ->
-          t (* Assertion that the condition always holds *)
-      | Interrupt, Interrupt -> Interrupt
-    (* End *)
-
-    (* Begin ControlFlowFromStmt *)
+    type abs_config =
+      | Abs_Throwing
+        (* evaluation of a statement yielded an a thrown exception. *)
+      | Abs_Returning (* evaluation of a return statement completed normally. *)
+      | Abs_Continuing
+        (* evaluation of a non-return statement completed normally. *)
+      | Abs_DynError
```

Comment:
Do we need to distinguish between `Abs_Throwing` and `Abs_DynError`? As far as I can see the checks always pair the two together